### PR TITLE
Allow all formats of dired listings

### PR DESCRIPTION
**Pull request by clemera**
Hi, thanks for this nice project! Dired also accepts custom file listings when a list is passed as argument, I added the ability to save those custom listings, too. You can test this by using something like:

    (dired
     '("*my-listing*" "~/.emacs.d" "~/.config"))

I also added storing history for dired buffers created by processes like `find-dired` or `find-grep-dired`.

### CODE REVIEW

**Comment by clemera** [comment]
I have found some issue with loading the history from disk, sometimes the `read` fails with the now present text properties but I haven't figured out why, yet.

**Comment by clemera** [comment]
Okay, fixed!

**Comment by vifon** [comment]
Supporting these "virtual" dired buffers wasn't initially in the scope of the project to keep things simple, though I'm not strictly against adding such support, though adding it as an opt-in feature would be welcome (controlled with a new `defcustom` variable).

As it currently is, I find your code hard to follow.  Would it be possible to break down these large `let` blocks into smaller functions?  Additionally I'm concerned about the calls such as `(current-buffer)` as the `dired-recent-path-save` function becomes much more stateful.  Did you check whether it still works with the optional argument passed?

**Comment by clemera** [comment]
Thanks for your feedback!

> adding it as an opt-in feature would be welcome (controlled with a new defcustom variable).

I have added the option `dired-recent-add-virtual-listings` to control this.

> Would it be possible to break down these large let blocks into smaller functions?

I have refactored the code to make it more readable.

>  Additionally I'm concerned about the calls such as (current-buffer) as the dired-recent-path-save function becomes much more stateful. Did you check whether it still works with the optional argument passed?

In master the state is captured by `default-directory` which can be wrong for virtual buffers when called from  `dired-mode-hook`. As the function is primarily intended to be called from the hook it might be okay to depend on that state. Users can now still pass any path to add one manually to the history as before.

**Comment by vifon** [comment]
Thanks for your patience. Looks good!

**Comment by vifon** [comment]
Merged, thank you!

**Comment by clemera** [comment]
Great, thanks!
